src/utils/networkUtils.ts

```diff
@@ -0,0 +1,183 @@
+  return FILTERABLE_RESOURCE_TYPES_SET.has(value as FilterableResourceType);
+}
+
+export function sanitizeRequestTypeFilter(
+  requestType?: string | string[] | null,
+): FilterableResourceType | FilterableResourceType[] | undefined {
+  if (requestType === undefined || requestType === null) {
+    return undefined;
+  }
+
+  const values = Array.isArray(requestType) ? requestType : [requestType];
+  const sanitized = values.filter(isFilterableResourceType);
+
+  if (!sanitized.length) {
+    return undefined;
+  }
+
+  return Array.isArray(requestType) ? sanitized : sanitized[0];
+}
+
+export function filterNetworkRequests(
+  requests: readonly HTTPRequest[],
+  requestType?: FilterableResourceType | FilterableResourceType[],
+): readonly HTTPRequest[] {
+  if (!requestType) {
+    return requests;
+  }
+
+  const normalizedTypes = new Set<FilterableResourceType>(
+    Array.isArray(requestType) ? requestType : [requestType],
+  );
+
+  if (!normalizedTypes.size) {
+    return requests;
+  }
+
+  return requests.filter(request => {
+    const type = request.resourceType();
+    if (!isFilterableResourceType(type)) {
+      return false;
+    }
+    return normalizedTypes.has(type);
+  });
+}
+
+export function paginateNetworkRequests(
+  requests: readonly HTTPRequest[],
+  options?: NetworkRequestsListingOptions,
+): NetworkRequestsListingResult {
+  const sanitizedOptions = options ?? {};
+  const filteredRequests = filterNetworkRequests(
+    requests,
+    sanitizedOptions.requestType,
+  );
+  const total = filteredRequests.length;
+
+  const hasPaginationOptions = hasPagination(sanitizedOptions);
+
+  if (!hasPaginationOptions) {
+    return {
+      requests: filteredRequests,
+      nextPageToken: undefined,
+      previousPageToken: undefined,
+      startIndex: 0,
+      endIndex: total,
+      invalidToken: false,
+      total,
+      appliedRequestType: sanitizedOptions.requestType,
+    };
+  }
+
+  const pageSize = validatePageSize(sanitizedOptions.pageSize, total);
+  const {startIndex, invalidToken} = resolveStartIndex(
+    sanitizedOptions.pageToken,
+    total,
+  );
+
+  const pageRequests = filteredRequests.slice(
+    startIndex,
+    startIndex + pageSize,
+  );
+  const endIndex = startIndex + pageRequests.length;
+
+  const nextPageToken = endIndex < total ? String(endIndex) : undefined;
+  const previousPageToken =
+    startIndex > 0 ? String(Math.max(startIndex - pageSize, 0)) : undefined;
+
+  return {
+    requests: pageRequests,
+    nextPageToken,
+    previousPageToken,
+    startIndex,
+    endIndex,
+    invalidToken,
+    total,
+    appliedRequestType: sanitizedOptions.requestType,
+  };
+}
+
+function hasPagination(options: NetworkRequestsListingOptions): boolean {
+  return (
+    options.pageSize !== undefined ||
+    (options.pageToken !== undefined && options.pageToken !== null)
+  );
+}
+
+function validatePageSize(pageSize: number | undefined, total: number): number {
```

Comment:
Would not the JSON schema already validate the page size? If not,  can we update the JSON schema.